streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java

```diff
@@ -298,14 +366,20 @@ public void evictWhile(final Supplier<Boolean> predicate,
 
             // predicate being true means we read one record, call the callback, and then remove it
             while (next != null && predicate.get()) {
+                if (next.getKey().time != minTimestamp) {
```

Comment:
This is one of the bugs that this PR fixes. The correct operation of this class depends on maintaining `minTimestamp` on every mutation. If a future code change breaks that contract, it would re-introduce a variant of our bug, and it would be just as subtle as it was this time. This check safeguards against that eventuality, by checking on our invariant at the spot we depend on it.

Reply:
Ack. So we don't need it, but we want it :)